polyglot/piranha/visualize_rules_graph.py

```diff
@@ -0,0 +1,234 @@
+        They get an additional label `(Cleanup_Rule)`.
+
+    Nodes without a group are added to a different set.
+    """
+    for rule_toml in rules_toml_dict['rules']:
+        rule_name: str = sanitize_name(rule_toml['name'])
+        if 'groups' in rule_toml:
+            rule_groups = rule_toml['groups']
+            collect_node_for_group_rule(rule_name, rule_groups)
+        else:  # rule without group
+            orphan_nodes.add(rule_name)
+
+
+def collect_node_for_group_rule(rule_name: str, rule_groups: 'list[str]'):
+    for group_name in rule_groups:
+        group_name = sanitize_name(group_name)
+        if group_name != 'Cleanup_Rule':
+            if group_name in rules_by_group_dict:
+                rules_by_group_dict[group_name].append(rule_name)
+            else:
+                rules_by_group_dict[group_name] = [rule_name]
+        else:  # we don't want to group 'Cleanup Rule's under the same shape
+            cleanup_rules.add(rule_name)
+
+
+def sanitize_name(s: str) -> str:
+    """Graphviz does not like names with spaces. Converts spaces to '_'."""
+    s = re.sub(r"\s+", '_', s)
+    return s
+
+
+def collect_edges(edges_toml_dict):
+    """
+    Groups outgoing edges by rule/group.
+
+    All names are sanitized to replace empty spaces by `_`.
+    """
+    for edge_toml in edges_toml_dict['edges']:
+        from_node = sanitize_name(edge_toml['from'])
+        to_nodes: 'list[str]' = edge_toml['to']
+        scope = sanitize_name(edge_toml['scope'])
+        for to_node in to_nodes:
+            to_node = sanitize_name(to_node)
+            edge = Edge(to=to_node, scope=scope)
+            if from_node in outgoing_edges_by_node:
+                outgoing_edges_by_node[from_node].append(edge)
+            else:
+                outgoing_edges_by_node[from_node] = [edge]
+
+
+def group_rule_names_label(rule_names: 'list[str]') -> 'list[str]':
+    return [inline_name(rule_name) for rule_name in rule_names]
+
+
+def inline_name(rule_name: str) -> str:
+    """Should be called for rules under a group."""
+    if rule_name in cleanup_rules:
+        return f'{rule_name} (Cleanup Rule)'
+    else:
+        return rule_name
+
+
+def initialize_graph() -> graphviz.Digraph:
+    graph_attr = {
+        'label': str(args.title),
+        'labelloc': 't',
+        'fontsize': '30'
+    }
+    graph = graphviz.Digraph(filename=output_file_path, graph_attr=graph_attr)
+    graph.format = 'svg'
+    return graph
+
+
+def generate_graph_nodes():
+    generate_non_orphan_nodes()
+    generate_orphan_nodes()
+
+
+def generate_non_orphan_nodes():
+    for rule_name in all_non_orphan_rule_names():
+        if rule_name in rules_by_group_dict:
+            generate_node_for_group_rule(rule_name)
+        else:  # a rule not in a group
+            generate_node_for_non_group_rule(rule_name)
+
+        added_nodes.add(rule_name)
+
+
+def all_non_orphan_rule_names() -> 'list[str]':
+    node_names_with_only_outgoing_edges = [
```

Comment:
is this the [set difference](https://www.geeksforgeeks.org/python-set-difference/) between the keys of two dicts?

Reply:
Nice catch. I changed it to use the set difference operator